lib/logdna/client.rb

```diff
@@ -1,164 +1,130 @@
       @side_messages = []
 
       @lock = Mutex.new
-      @task = nil
-
-      # NOTE: the byte limit only affects the message, not the entire message_hash
-      @actual_byte_limit = opts[:flushbyte] ||= Resources::FLUSH_BYTE_LIMIT
-      @actual_flush_interval = opts[:flushtime] ||= Resources::FLUSH_INTERVAL
-
-      @@request = request
-    end
-
-    def encode_message(msg)
-      msg = msg.to_s unless msg.instance_of? String
+      @flush_limit = opts[:flush_size] ? opts[:flush_size] : Resources::FLUSH_BYTE_LIMIT
+      @flush_interval = opts[:flush_interval] ? opts[:flush_interval] : Resources::FLUSH_INTERVAL
+      @flush_scheduled = false
+      @exception_flag = false
 
-      begin
-          msg = msg.encode("UTF-8")
-      rescue Encoding::UndefinedConversionError => e
-        # NOTE: should this be raised or handled silently?
-        # raise e
-      end
-      msg
+      @request = request
+      @retry_timeout = opts[:retry_timeout] ? opts[:retry_timeout] : Resources::RETRY_TIMEOUT
     end
 
-    def message_hash(msg, opts={})
-      obj = {
+    def process_message(msg, opts = {})
+      processed_message = {
         line: msg,
         app: opts[:app],
         level: opts[:level],
         env: opts[:env],
         meta: opts[:meta],
         timestamp: Time.now.to_i,
       }
-      obj.delete(:meta) if obj[:meta].nil?
-      obj
-    end
-
-    def create_flush_task
-      return @task unless @task.nil? or !@task.running?
-
-      t = Concurrent::TimerTask.new(execution_interval: @actual_flush_interval, timeout_interval: Resources::TIMER_OUT) do |task|
-        if @messages.any?
-          # keep running if there are queued messages, but don't flush
-          # because the buffer is being flushed due to being over the limit
-          unless @buffer_over_limit
-            flush()
-          end
-        else
-          # no messages means we can kill the task
-          task.kill
-        end
-      end
-      t.execute
-    end
-
-    def check_side_buffer
-      return if @side_buffer.size == 0
-
-      @buffer.write(@side_buffer.string)
-      @side_buffer.truncate(0)
-      queued_side_messages = @side_messages
-      @side_messages = []
-      queued_side_messages.each { |message_hash_obj| @messages.push(message_hash_obj) }
+      processed_message.delete(:meta) if processed_message[:meta].nil?
+      processed_message
     end
 
-
-    # this should always be running synchronously within this thread
-    def buffer(msg, opts)
-      buffer_size = write_to_buffer(msg, opts)
-      unless buffer_size.nil?
-        process_buffer(buffer_size)
-      end
+    def schedule_flush
+      @flush_scheduled = true
+      start_timer = lambda {
+        sleep(@exception_flag ? @retry_timeout : @flush_interval)
+        flush
+      }
+      thread = Thread.new { start_timer }
+      thread.join
     end
 
     def write_to_buffer(msg, opts)
-      return if msg.nil?
-      msg = encode_message(msg)
+      if @lock.try_lock
```

Comment:
So the way the lock logic has changed makes me a bit nervous. Before we just wrapped in a `synchronize` block so we know we released in the right place. Now there's odd coordination going on, and I'm not sure it's correct, like you're releasing the lock before we flush (line 62) - is that right? If you're concerned about blocking what about just checking for `if @lock.locked?` and adding to `@side_messages` in that case, otherwise doing this logic in a `synchronize` block?